// @babel/helper-import-to-platform-api/lib/platforms-support.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getSupport;
var _helperCompilationTargets = require("@babel/helper-compilation-targets");
const _excluded = ["node"];
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function isEmpty(obj) {
  return Object.keys(obj).length === 0;
}
const isRequiredOptions = {
  compatData: {
    webIMR: {
      chrome: "105.0.0",
      edge: "105.0.0",
      firefox: "106.0.0",
      opera: "91.0.0",
      safari: "16.4.0",
      opera_mobile: "72.0.0",
      ios: "16.4.0",
      samsung: "20.0",
      deno: "1.24.0"
    },
    nodeIMR: {
      node: "20.6.0"
    },
    nodeFSP: {
      node: "10.0.0"
    }
  }
};
const SUPPORT_CACHE = new WeakMap();
function getSupport(targets) {
  if (SUPPORT_CACHE.has(targets)) return SUPPORT_CACHE.get(targets);
  const {
      node: nodeTarget
    } = targets,
    webTargets = _objectWithoutPropertiesLoose(targets, _excluded);
  const emptyNodeTarget = nodeTarget == null;
  const emptyWebTargets = isEmpty(webTargets);
  const needsNodeSupport = !emptyNodeTarget || emptyWebTargets;
  const needsWebSupport = !emptyWebTargets || emptyNodeTarget;
  const webSupportsIMR = !emptyWebTargets && !(0, _helperCompilationTargets.isRequired)("webIMR", webTargets, isRequiredOptions);
  const nodeSupportsIMR = !emptyNodeTarget && !(0, _helperCompilationTargets.isRequired)("nodeIMR", {
    node: nodeTarget
  }, isRequiredOptions);
  const nodeSupportsFsPromises = !emptyNodeTarget && !(0, _helperCompilationTargets.isRequired)("nodeFSP", {
    node: nodeTarget
  }, isRequiredOptions);
  const result = {
    needsNodeSupport,
    needsWebSupport,
    nodeSupportsIMR,
    webSupportsIMR,
    nodeSupportsFsPromises
  };
  SUPPORT_CACHE.set(targets, result);
  return result;
}
